backend/rules/attackpath/PrivateVmAdmin.go

```diff
@@ -0,0 +1,147 @@
+		WITH a, e, instance_group_ids, public_elbv2_ids, collect(role.arn) as admin_roles, collect(role.admin_reason) as admin_reasons
+		WITH a, e, instance_group_ids, public_elbv2_ids, admin_roles, admin_reasons,
+		(e.publicipaddress IS NULL OR size(instance_group_ids) = 0) AND size(public_elbv2_ids) = 0 as private,
+		(size(admin_roles) > 0) as is_admin
+		RETURN e.id as resource_id,
+		'EC2Instance' as resource_type,
+		a.id as account_id,
+		CASE
+			WHEN private AND is_admin THEN 'failed'
+			ELSE 'passed'
+		END as status,
+		CASE
+			WHEN is_admin THEN (
+				'The instance is effectively an admin in the account because of: ' + admin_reasons[0] + '.'
+			)
+			ELSE 'The instance was not detected as effectively an admin in the account.'
+		END as context`,
+		nil,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []types.Result
+	for records.Next() {
+		record := records.Record()
+		resourceID, _ := record.Get("resource_id")
+		resourceIDStr, ok := resourceID.(string)
+		if !ok {
+			return nil, fmt.Errorf("resource_id %v should be of type string", resourceID)
+		}
+		resourceType, _ := record.Get("resource_type")
+		resourceTypeStr, ok := resourceType.(string)
+		if !ok {
+			return nil, fmt.Errorf("resource_type %v should be of type string", resourceType)
+		}
+		accountID, _ := record.Get("account_id")
+		accountIDStr, ok := accountID.(string)
+		if !ok {
+			return nil, fmt.Errorf("account_id %v should be of type string", accountID)
+		}
+		status, _ := record.Get("status")
+		statusStr, ok := status.(string)
+		if !ok {
+			return nil, fmt.Errorf("status %v should be of type string", status)
+		}
+		context, _ := record.Get("context")
+		contextStr, ok := context.(string)
+		if !ok {
+			return nil, fmt.Errorf("context %v should be of type string", context)
+		}
+		results = append(results, types.Result{
+			ResourceID:   resourceIDStr,
+			ResourceType: resourceTypeStr,
+			AccountID:    accountIDStr,
+			Status:       statusStr,
+			Context:      contextStr,
+		})
+	}
+	return results, nil
+}
+
+func (PrivateVmAdmin) ProduceRuleGraph(tx neo4j.Transaction, resourceId string) (types.GraphPathResult, error) {
+	var params = make(map[string]interface{})
+	params["InstanceId"] = resourceId
+	records, err := tx.Run(
+		`MATCH (a:AWSAccount{inscope: true})-[:RESOURCE]->(e:EC2Instance{id: $InstanceId})
+		OPTIONAL MATCH
+			directPath=
+			(e)-[:MEMBER_OF_EC2_SECURITY_GROUP|NETWORK_INTERFACE*..2]->(instance_group:EC2SecurityGroup)
+			<-[:MEMBER_OF_EC2_SECURITY_GROUP]-(:IpPermissionInbound)
+			<-[:MEMBER_OF_IP_RULE]-(:IpRange)
+		WITH e, collect(directPath) as directPaths
+		OPTIONAL MATCH
+			adminRolePath=
+			(e)-[:STS_ASSUME_ROLE_ALLOW]->(role:AWSRole{is_admin: True})
+		WITH e, directPaths, collect(adminRolePath) as adminRolePaths
+		WITH directPaths + adminRolePaths AS paths
+		RETURN paths`,
+		params)
+	if err != nil {
+		return types.GraphPathResult{}, err
+	}
+
+	// Parsed out graph from the query.
```

Comment:
ProcessGraphPathResult, GraphStartNodeCheck, and CompressPaths are query agnostic. Don't put them in the rule spec. Put them in GetRuleGraph